### ISSUE TITLE
Discussion about normalizations in representers

### ISSUE THREAD

**Issue by mpizenberg** [ceddlyburge/elm-representer#3]
I've had some chat with Jeremy, I'll paste screenshots here cause there are too many things to copy-paste ^^

![Screenshot_2021-04-10 Slack Threads Exercism Team 1 new item](https://user-images.githubusercontent.com/2905865/114275957-d05a2180-9a24-11eb-85ec-542e93d946a0.png)
![Screenshot_2021-04-10 Slack Threads Exercism Team 1 new item(1)](https://user-images.githubusercontent.com/2905865/114275960-d51ed580-9a24-11eb-9909-8cf6371948d6.png)
![Screenshot_2021-04-10 Slack Threads Exercism Team 1 new item(2)](https://user-images.githubusercontent.com/2905865/114275964-d94af300-9a24-11eb-858b-6d50f7ad8938.png)

**Comment by ceddlyburge** [comment]
Hi Matthieu, it sounds like out outcome of this was that we should normalize within a scope, is that right?

That will require some changes to the normalization, and raises some questions for the mapping file / dictionary (name is no longer enough, it would need to be name and scope). It will definitely reduce the number of "unique" solutions that require mentoring tho ...

**Comment by mpizenberg** [comment]
Actually, we've continued our discussion afterwards and I'm we should keep
the relation between original and normalized names reversible. I'll answer
with more details tonight.
Cheers

**Comment by mpizenberg** [comment]
Here is a follow up discussion on the subject of representers: https://github.com/exercism/ruby-representer/pull/18#issue-612979240

Concretely the two important things to note are in my opinion:
1. Representers could return `notes.md` along with the `representation.txt` to inform a mentor on what normalisations have occurred, and therefore what not to give feedback on (like function names for example).
2. For the time being, we keep normalized names equivalent to original names, *"all identical names must be replaced with the same placeholder, irrespective of scope"*, like you currently do. I've introduced my thoughts about "representation levels" in the issue linked above. I think that would be pretty neat but let's wait for after v3 launch and keep it simple for the time being.

**Comment by ceddlyburge** [comment]
Hi Matthieu.

As I understand it, mentors cannot usefully give feedback on formatting / whitespace, comments or identifier names with the current normalization.

If we normalize identifer names within a scope (which would probably reduce the number of "unique" solutions a lot), I don't think this changes, so it seems sensible to go ahead with it. The mapping.json file no longer makes total sense, but I don't think I understand why this matters (mentors can work with the unnormalized file without problems, which is probably a lot easier to understand than the normalized one, or they could work with the normalized file if they are feeling brave).

As an aside, have you come across the Unison language. Definitions are identified by their content in Unison, which is a super interesting idea, and vaguely relevant to normalization / uniquely idenitifier semantically identical code.

**Comment by mpizenberg** [comment]
> As I understand it, mentors cannot usefully give feedback on formatting / whitespace, comments or identifier names with the current normalization.

In the current state, mentors cannot cannot comment identifier names but they can use identifiers to uniquely identify variables. If we go ahead with normalization within scopes, this is not the case anymore. Let's look at this for example.
```elm
add2 x = x + 2
add3 y = y + 3
```
Currently, `x` and `y` will receive different identifiers, so a mentoring note could refer to `y` directly in a non-ambiguous way. If we normalize variables with local scope, the normalization will be equivalent to
```elm
add2 x = x + 2
add3 x = x + 3
```
As such, mentoring notes would need to refer to "`x` in the add3 function" or something similar. I think that would be fine, if the `notes.md` file coming with the representations clearly states the types of normalization done and the mentor creating the feedback is aware of this. In practice, those automated feedbacks will only be created by "mentor veterans" that have contributed a lot to exercism so I think it should be safe to do that optimization. I think if you'd like to go for it, you should! Maybe, if we want to make that easily removable, if that causes issues with the exercism platform, just try to do it in a modular way. Something like an additional normalization pass maybe, that we could easily disable.

> As an aside, have you come across the Unison language

Yeah :) I've watched a few talks about it. Definitely feels like a really interesting approach and I'll try to follow progress on that project regularly out of curiosity.

**Comment by mpizenberg** [comment]
> As such, mentoring notes would need to refer to "x in the add3 function" or something similar.

Actually let me take that back! It will not be possible anymore to reference that variable in an explanation if the solution was using a different variable name (like y) because "x in the add3 function" simply does not exist. So the only way this _could work_ would be with an intermediate layer as the one described by Jeremy in the PR I linked above.

**Comment by mpizenberg** [comment]
One thing I'd love to have concrete numbers on would be the actual improvements and number unique representations for any given exercism exercise. Maybe if you're up for it, try discussing this with Jeremy to see if you could get access to data for the elm track to support that research. Like so, we could figure out tradeoffs between normalization improvements and explainability for feedback on concrete results.

**Comment by ceddlyburge** [comment]
So the idea is that the system will track references to identifiers somehow, so that a mentor can leave feedback using the idenfitiers in a solution, the system will store these as the normalized identifers, and then unnormalize them when applying them to a solution with different names?

Being as it makes no sense to leave feedback on the names themselves, the identifiers are then only being used to describe which bit of code the mentor is talking about. I wonder whether using line numbers (or something similar) instead would get around the problem? Is this what you are talking about with the layered idea? As things like sorting and format and whitespace will all affect the line numbers. Potentially mentors could even highlight code, and the system can map the start and end of the code snippet between normalized and unnormalized solutions.

**Comment by mpizenberg** [comment]
> So the idea is that the system will track references to identifiers somehow

Yep, that is what the `mapping.json` is for, at least that's my understanding: https://github.com/exercism/docs/blob/main/building/tooling/representers/interface.md

> I wonder whether using line numbers (or something similar) instead would get around the problem?

Well in javascript world, I believe that's what source maps are for, so if we push this idea even further, we could use source maps as a reference for feedbacks. That would mean a lot more work for every track and for the website UX so I don't think that is happening soon :)

**Comment by ceddlyburge** [comment]
So at the moment mentors are *required* to only reference the code by the identifiers so that the system can apply the same feedback to multiple solutions. Interesting, and makes sense when you think about it.

But it does sound like the overall upshot is that we leave the normalization as it is and don't take account of scope, and maybe in the future there will be some code of source map solution.

**Comment by mpizenberg** [comment]
> mentors are required to only reference the code by the identifiers

Either that, or they can use the names of the solution but the 1-to-1 mapping.json enables translating that automatically for other solutions. To be honest I haven't tried that part of exercism.lol yet and don't know how it's implemented.
